e2e/tests/dump-restore-test.rs

```diff
@@ -181,61 +277,151 @@ fn dump_db(container: &Container<Cli, GenericImage>, db: &str, username: &str) -
+}
 
-    let container = run_postgres(&docker, "db1", "postgres");
-    psql_cmd(
+/// Runs the pre-restore sql script in the container
+fn pre_restore(container: &PostgresContainer) {
+    psql_file(
         &container,
-        "db1",
+        "db",
         "postgres",
-        "create user jill; grant all on database db1 to jill; grant postgres to jill;", // todo: jill should not need postgres role
+        Path::new("/scripts/pre-restore.sql"),
     );
-    psql_cmd(
+}
+
+/// Runs the post-restore sql script in the container
+fn post_restore(container: &PostgresContainer) {
+    psql_file(
         &container,
-        "db1",
-        "postgres", // todo: use jill for this
-        "create extension if not exists timescaledb; create extension if not exists promscale; select public.timescaledb_pre_restore();"
+        "db",
+        "postgres", // todo: use bob for this
+        Path::new("/scripts/post-restore.sql"),
     );
-    restore_db(&container, "db1", "postgres", dump);
-    psql_cmd(
+}
+
+/// Runs the post-snapshot sql script in the container
+fn post_snapshot(container: &PostgresContainer) {
+    psql_file(
         &container,
-        "db1",
-        "postgres", // todo: use jill for this
-        "select public.timescaledb_post_restore();",
+        "db",
+        "bob",
+        Path::new("/scripts/post-snapshot.sql"),
     );
-    psql_cmd(&container, "db1", "postgres", "grant prom_admin to jill;");
-    let snapshot1 = snapshot_db(&container, "db1", "jill");
-    container.stop();
-    container.rm();
+}
+
+/// Tests the process of dumping and restoring a database using pg_dump
+///
+/// 1. create a postgres docker container
+/// 2. create metrics, series, samples, exemplars, traces in the database
+/// 3. snapshot the database
+/// 4. use pg_dump to create a logical sql dump
+/// 5. destroy the docker container and create a new one
+/// 6. restore into the new database
+/// 7. snapshot the database
+/// 8. create more data to make sure things are still wired up correctly
+/// 9. compare the two snapshots. they should be equal
+#[test]
+fn dump_restore_test() {
+    let docker = clients::Cli::default();
+    let postgres_image = postgres_image();
+    let volumes = vec![(concat!(env!("CARGO_MANIFEST_DIR"), "/scripts"), "/scripts")];
+
+    let dir = &env::temp_dir().join("promscale-dump-restore-test");
+    println!("temp dir at: {}", dir.to_str().unwrap());
+    if dir.exists() {
+        fs::remove_dir_all(dir).expect("failed to remove temp dir");
+    }
+    fs::create_dir_all(dir).expect("failed to create temp dir");
+    let dump = dir.join("dump.sql");
+
+    // create the first container, load it with data, snapshot it, and dump it
+    let snapshot0 = {
+        let container = run_postgres(
+            &docker,
+            postgres_image.as_str(),
+            "db",
+            "postgres",
+            Some(&volumes),
+        );
+        pre_dump(&container);
+        let snapshot0 = snapshot_db(
```

Comment:
not sure if this is a rust thing but I usually don't like to shadow variables like this. Snapshot0 appears in 2 places and are different vars.

Reply:
does it make you like it more or less that this is an anonymous block, and the result of the block is `snapshot0`? Yes, they are 2 different variables, but they contain the same values.

I'll try to break these out to separate functions. Maybe my rust-fu is strong enough now to deal with that

Reply:
resolved in [9263761](https://github.com/timescale/promscale_extension/pull/185/commits/92637615120da4f7db03e4c6c6adb2e08e5f66ca)